sharding/review/casper_v2.1.md

```diff
@@ -0,0 +1,362 @@
+    cofactor = 39
+    STANDARD_COMMITTEE_SIZE = MAX_VALIDATOR_COUNT // SHARD_COUNT
+    # If there are not enough validators to fill a minimally
+    # sized committee at every height, skip some heights
+    if validator_count < EPOCH_LENGTH * MIN_COMMITTEE_SIZE:
+        height_count = validator_count // MIN_COMMITTEE_SIZE or 1
+        heights = [(i * cofactor) % EPOCH_LENGTH
+                   for i in range(height_count)]
+    # If there are enough validators, fill all the heights
+    else:
+        height_count = EPOCH_LENGTH
+        heights = list(range(EPOCH_LENGTH))
+            
+    filled = 0
+    for i in range(EPOCH_LENGTH - 1):
+        if not i in heights:
+            height_cutoffs.append(height_cutoffs[-1])
+        else:
+            filled += 1
+            height_cutoffs.append(filled * validator_count // height_count)
+    height_cutoffs.append(validator_count)
+    
+    # For the validators assigned to each height, split them up
+    # into committees for different shards. Do not assign the
+    # last END_EPOCH_GRACE_PERIOD heights in an epoch to any shards.
+    shard_cutoffs = [0]
+    for i in range(EPOCH_LENGTH - END_EPOCH_GRACE_PERIOD):
+        size = height_cutoffs[i+1] - height_cutoffs[i]
+        shards = (size + STANDARD_COMMITTEE_SIZE - 1) // STANDARD_COMMITTEE_SIZE
+        pre = shard_cutoffs[-1]
+        for j in range(1, shards+1):
+            shard_cutoffs.append(pre + size * j // shards)
+        
+    return height_cutoffs, shard_cutoffs
+```
+
+This splits up the validator set into groups for each height, and within each height into groups for different shards. Note that if the validator set is too small to support a committee at every height, some heights are simply left "inactive".
+
+Here's a depiction of what is going on with reduced parameters for simplicity of exposition:
+
+![](http://vitalik.ca/files/ShuffleAndAssign.png)
+
+The `current_shuffling` is recalculated at the start of a dynasty transition.
+
+### Per-block processing
+
+The per-block processing is broken down into three parts.
+
+#### Checking partial crosslink records
+
+A block can have 0 or more `AggregateVote` objects, where each `AggregateVote` object has the following fields:
+
+```python
+fields = {
+    'height': 'int16',
+    'parent_hash': 'hash32',
+    'checkpoint_hash': 'hash32',
+    'shard_id': 'int16',
+    'shard_block_hash': 'hash32',
+    'attester_bitfield': 'bytes',
+    'aggregate_sig': ['int256']
+}
+```
+
+For each one of these votes:
+
+* Verify that `height <= slot_number`
+* Use `get_cutoffs` above to get the index cutoffs for the heights and the shards.
+    * If `height_in_epoch < EPOCH_LENGTH - END_EPOCH_GRACE_PERIOD`, verify that `si = (shard_id - next_shard) % SHARD_COUNT` is a valid shard index for that height (ie. check `height_cutoffs[height_in_epoch] <= shard_cutoffs[si] < height_cutoffs[height_in_epoch + 1])`).
+    * Otherwise, verify `shard_id == 65535` and `shard_block_hash == 0`.
```

Comment:
`shard_id == 65535` implies that `MAX_SHARD_COUNT` is `2^16`. Why not set `MAX_SHARD_COUNT`to `2^15`, where `MAX_VALIDATOR_COUNT / MIN_COMMITTEE_SIZE = 2^15`?

Reply:
The point is to have an "out of the way" shard ID that doesn't represent any shard.